lab3van/util-linux/disk-utils/mkfs.8
<<<<<<< HEAD
.\" -*- nroff -*-
=======
>>>>>>> master-vanilla
.TH MKFS 8 "June 2011" "util-linux" "System Administration"
.SH NAME
mkfs \- build a Linux filesystem
.SH SYNOPSIS
<<<<<<< HEAD
.SH SYNOPSIS
.B mkfs
.RI [ options ]
.RB [ \-t
.IR "type fs-options" ] " device " [ size ]
.SH DESCRIPTION
=======
.B mkfs
[options]
.RB [ \-t
.IR type "] [" fs-options ] " device " [ size ]
.SH DESCRIPTION
.B This mkfs frontend is deprecated in favour of filesystem specific mkfs.<type> utils.
.PP
>>>>>>> master-vanilla
.B mkfs
is used to build a Linux filesystem on a device, usually
a hard disk partition.  The
.I device
argument is either the device name (e.g.
.IR /dev/hda1 ,
.IR /dev/sdb2 ),
or a regular file that shall contain the filesystem.  The
.I size
argument is the number of blocks to be used for the filesystem.
.PP
The exit code returned by
.B mkfs
is 0 on success and 1 on failure.
.PP
In actuality,
.B mkfs
is simply a front-end for the various filesystem builders
(\fBmkfs.\fIfstype\fR)
available under Linux.
<<<<<<< HEAD
The filesystem-specific builder is searched for in a number
of directories, like perhaps
.IR /sbin ,
.IR /sbin/fs ,
.IR /sbin/fs.d ,
.IR /etc/fs ,
.I /etc
(the precise list is defined at compile time but at least
contains
.I /sbin
and
.IR /sbin/fs ),
and finally in the directories
listed in the PATH environment variable.
=======
The filesystem-specific builder is searched for via your PATH
environment setting only.
>>>>>>> master-vanilla
Please see the filesystem-specific builder manual pages for
further details.
.SH OPTIONS
.TP
.BR \-t , " \-\-type " \fItype\fR
Specify the \fItype\fR of filesystem to be built.
If not specified, the default filesystem type
(currently ext2) is used.
.TP
.I fs-options
Filesystem-specific options to be passed to the real filesystem builder.
<<<<<<< HEAD
Although not guaranteed, the following options are supported
by most filesystem builders.
=======
>>>>>>> master-vanilla
.TP
.BR \-V , " \-\-verbose"
Produce verbose output, including all filesystem-specific commands
that are executed.
Specifying this option more than once inhibits execution of any
filesystem-specific commands.
This is really only useful for testing.
.TP
.BR \-V , " \-\-version"
Display version information and exit.  (Option \fB\-V\fR will display
version information only when it is the only parameter, otherwise it
will work as \fB\-\-verbose\fR.)
.TP
.BR \-h , " \-\-help"
<<<<<<< HEAD
Display help and exit.
.SH BUGS
All generic options must precede and not be combined with
filesystem-specific options.
Some filesystem-specific programs do not support the
.B -V
(verbose) option, nor return meaningful exit codes.
Also, some filesystem-specific programs do not automatically
=======
Display help text and exit.
.SH BUGS
All generic options must precede and not be combined with
filesystem-specific options.
Some filesystem-specific programs do not automatically
>>>>>>> master-vanilla
detect the device size and require the
.I size
parameter to be specified.
.SH AUTHORS
David Engel (david@ods.com)
.br
<<<<<<< HEAD
Fred N. van Kempen (waltje@uwalt.nl.mugnet.org)
=======
Fred N.\& van Kempen (waltje@uwalt.nl.mugnet.org)
>>>>>>> master-vanilla
.br
Ron Sommeling (sommel@sci.kun.nl)
.br
The manual page was shamelessly adapted from Remy Card's version
for the ext2 filesystem.
.SH SEE ALSO
<<<<<<< HEAD
=======
.na
>>>>>>> master-vanilla
.BR fs (5),
.BR badblocks (8),
.BR fsck (8),
.BR mkdosfs (8),
.BR mke2fs (8),
.BR mkfs.bfs (8),
.BR mkfs.ext2 (8),
.BR mkfs.ext3 (8),
.BR mkfs.ext4 (8),
.BR mkfs.minix (8),
.BR mkfs.msdos (8),
.BR mkfs.vfat (8),
<<<<<<< HEAD
.BR mkfs.xfs (8),
.BR mkfs.xiafs (8)
=======
.BR mkfs.xfs (8)
.ad
>>>>>>> master-vanilla
.SH AVAILABILITY
The mkfs command is part of the util-linux package and is available from
ftp://ftp.kernel.org/pub/linux/utils/util-linux/.


debian/changelog
util-linux (2.27.160-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- roma <newangel88@mail.ru>  Thu, 24 Dec 2015 13:07:17 +0400


lab3van/util-linux/disk-utils/mkfs.bfs.c
/*
 *  mkfs.bfs - Create SCO BFS filesystem - aeb, 1999-09-07
 *
<<<<<<< HEAD
=======
 *	Usage: mkfs.bfs [-N nr-of-inodes] [-V volume-name] [-F fsname] device
>>>>>>> master-vanilla
 */

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

#include "blkdev.h"
#include "c.h"
<<<<<<< HEAD
#include "nls.h"
#include "strutils.h"
#include "xalloc.h"
=======
#include "closestream.h"
#include "nls.h"
#include "strutils.h"
#include "xalloc.h"
#include "bitops.h"
>>>>>>> master-vanilla

#define BFS_ROOT_INO		2
#define BFS_NAMELEN		14
#define BFS_BLOCKSIZE		512
#define BFS_SUPER_MAGIC		0x1badface

/* superblock - 512 bytes */
struct bfssb {
<<<<<<< HEAD
	unsigned int s_magic;
	unsigned int s_start;	/* byte offset of start of data */
	unsigned int s_end;	/* sizeof(slice)-1 */

	/* for recovery during compaction */
	int s_from, s_to;	/* src and dest block of current transfer */
	int s_backup_from, s_backup_to;
=======
	uint32_t s_magic;
	uint32_t s_start;	/* byte offset of start of data */
	uint32_t s_end;	/* sizeof(slice)-1 */

	/* for recovery during compaction */
	uint32_t s_from, s_to;	/* src and dest block of current transfer */
	int32_t s_backup_from, s_backup_to;
>>>>>>> master-vanilla

	/* labels - may well contain garbage */
	char s_fsname[6];
	char s_volume[6];
	char s_pad[472];
};

/* inode - 64 bytes */
struct bfsi {
<<<<<<< HEAD
	unsigned short i_ino;
	unsigned char i_pad1[2];
	unsigned long i_first_block;
	unsigned long i_last_block;
	unsigned long i_bytes_to_end;
	unsigned long i_type;	/* 1: file, 2: the unique dir */
	unsigned long i_mode;
	unsigned long i_uid, i_gid;
	unsigned long i_nlinks;
	unsigned long i_atime, i_mtime, i_ctime;
=======
	uint16_t i_ino;
	unsigned char i_pad1[2];
	uint32_t i_first_block;
	uint32_t i_last_block;
	uint32_t i_bytes_to_end;
	uint32_t i_type;	/* 1: file, 2: the unique dir */
	uint32_t i_mode;
	uint32_t i_uid, i_gid;
	uint32_t i_nlinks;
	uint32_t i_atime, i_mtime, i_ctime;
>>>>>>> master-vanilla
	unsigned char i_pad2[16];
};

#define BFS_DIR_TYPE	2

/* directory entry - 16 bytes */
struct bfsde {
<<<<<<< HEAD
	unsigned short d_ino;
=======
	uint16_t d_ino;
>>>>>>> master-vanilla
	char d_name[BFS_NAMELEN];
};

static void __attribute__ ((__noreturn__)) usage(FILE * out)
{
	fprintf(out,
		_("Usage: %s [options] device [block-count]\n"),
		program_invocation_short_name);
<<<<<<< HEAD
=======

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Make an SCO bfs filesystem.\n"), out);

>>>>>>> master-vanilla
	fprintf(out, _("\nOptions:\n"
		       " -N, --inodes=NUM    specify desired number of inodes\n"
		       " -V, --vname=NAME    specify volume name\n"
		       " -F, --fname=NAME    specify file system name\n"
		       " -v, --verbose       explain what is being done\n"
		       " -c                  this option is silently ignored\n"
		       " -l                  this option is silently ignored\n"
		       " -V, --version       output version information and exit\n"
		       "                     -V as version must be only option\n"
		       " -h, --help          display this help and exit\n\n"));

<<<<<<< HEAD
=======
	fprintf(out, USAGE_MAN_TAIL("mkfs.bfs(8)"));
>>>>>>> master-vanilla
	exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

static void __attribute__ ((__noreturn__)) print_version(void)
{
<<<<<<< HEAD
	printf(_("%s from %s\n"), program_invocation_short_name, PACKAGE_STRING);
=======
	printf(UTIL_LINUX_VERSION);
>>>>>>> master-vanilla
	exit(EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	char *device, *volume, *fsname;
	long inodes;
	unsigned long long total_blocks, ino_bytes, ino_blocks, data_blocks;
	unsigned long long user_specified_total_blocks = 0;
	int verbose = 0;
	int fd;
<<<<<<< HEAD
=======
	uint32_t first_block;
>>>>>>> master-vanilla
	struct bfssb sb;
	struct bfsi ri;
	struct bfsde de;
	struct stat statbuf;
	time_t now;
	int c, i, len;

	enum { VERSION_OPTION = CHAR_MAX + 1 };
	static const struct option longopts[] = {
		{"inodes", required_argument, NULL, 'N'},
		{"vname", required_argument, NULL, 'V'},
		{"fname", required_argument, NULL, 'F'},
		{"verbose", no_argument, NULL, 'v'},
		{"version", no_argument, NULL, VERSION_OPTION},
		{"help", no_argument, NULL, 'h'},
		{NULL, 0, NULL, 0}
	};

<<<<<<< HEAD
=======
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	atexit(close_stdout);

>>>>>>> master-vanilla
	if (argc < 2)
		usage(stderr);

	if (argc == 2 && !strcmp(argv[1], "-V"))
		print_version();

	volume = fsname = "      ";	/* is there a default? */
	inodes = 0;

	while ((c = getopt_long(argc, argv, "N:V:F:vhcl", longopts, NULL)) != -1) {
		switch (c) {
		case 'N':
			inodes = strtol_or_err(optarg, _("invalid number of inodes"));
			break;

		case 'V':
			len = strlen(optarg);
			if (len <= 0 || len > 6)
				errx(EXIT_FAILURE, _("volume name too long"));
			volume = xstrdup(optarg);
			break;

		case 'F':
			len = strlen(optarg);
			if (len <= 0 || len > 6)
				errx(EXIT_FAILURE, _("fsname name too long"));
			fsname = xstrdup(optarg);
			break;

		case 'v':
			verbose = 1;
			break;

		case 'c':
		case 'l':
			/* when called via mkfs we may get options c,l,v */
			break;

		case VERSION_OPTION:
			print_version();
		case 'h':
			usage(stdout);
		default:
			usage(stderr);
		}
	}

	if (optind == argc)
		usage(stderr);

	device = argv[optind++];

	if (stat(device, &statbuf) < 0)
<<<<<<< HEAD
		err(EXIT_FAILURE, _("cannot stat device %s"), device);

	if (!S_ISBLK(statbuf.st_mode))
		errx(EXIT_FAILURE, _("%s is not a block special device"), device);

	fd = open(device, O_RDWR | O_EXCL);
=======
		err(EXIT_FAILURE, _("stat of %s failed"), device);

	fd = open_blkdev_or_file(&statbuf, device, O_RDWR);
>>>>>>> master-vanilla
	if (fd < 0)
		err(EXIT_FAILURE, _("cannot open %s"), device);

	if (optind == argc - 1)
		user_specified_total_blocks =
<<<<<<< HEAD
			strtoll_or_err(argv[optind], _("invalid block-count"));
=======
			strtou64_or_err(argv[optind], _("invalid block-count"));
>>>>>>> master-vanilla
	else if (optind != argc)
		usage(stderr);

	if (blkdev_get_sectors(fd, &total_blocks) == -1) {
		if (!user_specified_total_blocks)
			err(EXIT_FAILURE, _("cannot get size of %s"), device);
		total_blocks = user_specified_total_blocks;
	} else if (user_specified_total_blocks) {
		if (user_specified_total_blocks > total_blocks)
			errx(EXIT_FAILURE,
			     _("blocks argument too large, max is %llu"),
			     total_blocks);
		total_blocks = user_specified_total_blocks;
	}

	if (!inodes) {
		/* pick some reasonable default */
		inodes = 8 * (total_blocks / 800);
		if (inodes < 48)
			inodes = 48;
		if (512 < inodes)
			inodes = 512;
	} else {
		/* believe the user */
		if (512 < inodes)
			errx(EXIT_FAILURE, _("too many inodes - max is 512"));
	}

	ino_bytes = inodes * sizeof(struct bfsi);
	ino_blocks = (ino_bytes + BFS_BLOCKSIZE - 1) / BFS_BLOCKSIZE;
	data_blocks = total_blocks - ino_blocks - 1;

<<<<<<< HEAD
	/* mimic the behaviour of SCO's mkfs - maybe this limit is needed */
=======
	/* mimic the behavior of SCO's mkfs - maybe this limit is needed */
>>>>>>> master-vanilla
	if (data_blocks < 32)
		errx(EXIT_FAILURE,
		     _("not enough space, need at least %llu blocks"),
		     ino_blocks + 33);

	memset(&sb, 0, sizeof(sb));
<<<<<<< HEAD
	sb.s_magic = BFS_SUPER_MAGIC;
	sb.s_start = ino_bytes + sizeof(struct bfssb);
	sb.s_end = total_blocks * BFS_BLOCKSIZE - 1;
=======
	sb.s_magic = cpu_to_le32(BFS_SUPER_MAGIC);
	sb.s_start = cpu_to_le32(ino_bytes + sizeof(struct bfssb));
	sb.s_end = cpu_to_le32(total_blocks * BFS_BLOCKSIZE - 1);
>>>>>>> master-vanilla
	sb.s_from = sb.s_to = sb.s_backup_from = sb.s_backup_to = -1;
	memcpy(sb.s_fsname, fsname, 6);
	memcpy(sb.s_volume, volume, 6);

	if (verbose) {
		fprintf(stderr, _("Device: %s\n"), device);
		fprintf(stderr, _("Volume: <%-6s>\n"), volume);
		fprintf(stderr, _("FSname: <%-6s>\n"), fsname);
		fprintf(stderr, _("BlockSize: %d\n"), BFS_BLOCKSIZE);
		if (ino_blocks == 1)
<<<<<<< HEAD
			fprintf(stderr, _("Inodes: %lu (in 1 block)\n"),
				inodes);
		else
			fprintf(stderr, _("Inodes: %lu (in %llu blocks)\n"),
				inodes, ino_blocks);
		fprintf(stderr, _("Blocks: %lld\n"), total_blocks);
		fprintf(stderr, _("Inode end: %d, Data end: %d\n"),
			sb.s_start - 1, sb.s_end);
=======
			fprintf(stderr, _("Inodes: %ld (in 1 block)\n"),
				inodes);
		else
			fprintf(stderr, _("Inodes: %ld (in %llu blocks)\n"),
				inodes, ino_blocks);
		fprintf(stderr, _("Blocks: %llu\n"), total_blocks);
		fprintf(stderr, _("Inode end: %d, Data end: %d\n"),
			le32_to_cpu(sb.s_start) - 1, le32_to_cpu(sb.s_end));
>>>>>>> master-vanilla
	}

	if (write(fd, &sb, sizeof(sb)) != sizeof(sb))
		err(EXIT_FAILURE, _("error writing superblock"));

	memset(&ri, 0, sizeof(ri));
<<<<<<< HEAD
	ri.i_ino = BFS_ROOT_INO;
	ri.i_first_block = 1 + ino_blocks;
	ri.i_last_block = ri.i_first_block +
	    (inodes * sizeof(de) - 1) / BFS_BLOCKSIZE;
	ri.i_bytes_to_end = ri.i_first_block * BFS_BLOCKSIZE
	    + 2 * sizeof(struct bfsde) - 1;
	ri.i_type = BFS_DIR_TYPE;
	ri.i_mode = S_IFDIR | 0755;	/* or just 0755 */
	ri.i_uid = 0;
	ri.i_gid = 1;			/* random */
	ri.i_nlinks = 2;
	time(&now);
	ri.i_atime = now;
	ri.i_mtime = now;
	ri.i_ctime = now;
=======
	ri.i_ino = cpu_to_le16(BFS_ROOT_INO);
	first_block = 1 + ino_blocks;
	ri.i_first_block = cpu_to_le32(first_block);
	ri.i_last_block = cpu_to_le32(first_block +
	    (inodes * sizeof(de) - 1) / BFS_BLOCKSIZE);
	ri.i_bytes_to_end = cpu_to_le32(first_block * BFS_BLOCKSIZE
	    + 2 * sizeof(struct bfsde) - 1);
	ri.i_type = cpu_to_le32(BFS_DIR_TYPE);
	ri.i_mode = cpu_to_le32(S_IFDIR | 0755);	/* or just 0755 */
	ri.i_uid = cpu_to_le32(0);
	ri.i_gid = cpu_to_le32(1);			/* random */
	ri.i_nlinks = 2;
	time(&now);
	ri.i_atime = cpu_to_le32(now);
	ri.i_mtime = cpu_to_le32(now);
	ri.i_ctime = cpu_to_le32(now);
>>>>>>> master-vanilla

	if (write(fd, &ri, sizeof(ri)) != sizeof(ri))
		err(EXIT_FAILURE, _("error writing root inode"));

	memset(&ri, 0, sizeof(ri));
	for (i = 1; i < inodes; i++)
		if (write(fd, &ri, sizeof(ri)) != sizeof(ri))
			err(EXIT_FAILURE, _("error writing inode"));

	if (lseek(fd, (1 + ino_blocks) * BFS_BLOCKSIZE, SEEK_SET) == -1)
		err(EXIT_FAILURE, _("seek error"));

	memset(&de, 0, sizeof(de));
<<<<<<< HEAD
	de.d_ino = BFS_ROOT_INO;
=======
	de.d_ino = cpu_to_le16(BFS_ROOT_INO);
>>>>>>> master-vanilla
	memcpy(de.d_name, ".", 1);
	if (write(fd, &de, sizeof(de)) != sizeof(de))
		err(EXIT_FAILURE, _("error writing . entry"));

	memcpy(de.d_name, "..", 2);
	if (write(fd, &de, sizeof(de)) != sizeof(de))
		err(EXIT_FAILURE, _("error writing .. entry"));

<<<<<<< HEAD
	if (close(fd) < 0)
=======
	if (close_fd(fd) != 0)
>>>>>>> master-vanilla
		err(EXIT_FAILURE, _("error closing %s"), device);

	return EXIT_SUCCESS;
}
